plugins/ccgx/fu-ccgx-hpi.c

```diff
@@ -0,0 +1,1633 @@
+	CyPDResp hpi_event = 0;
+
+	g_return_val_if_fail (hpi_handle != NULL, FALSE);
+	g_return_val_if_fail (data != NULL, FALSE);
+	g_return_val_if_fail (size > 0, FALSE);
+
+	(void)fu_ccgx_hpi_clear_all_event(device,hpi_handle,HPI_CMD_COMMAND_CLEAR_EVENT_TIME_MS,error);
+
+	if (!fu_ccgx_hpi_read_flash(device, hpi_handle, row_num, error)) {
+		g_prefix_error (error, "read flash cmd error: ");
+		return FALSE;
+	}
+
+	/* wait until flash is reaad */
+	g_usleep (HPI_CMD_FLASH_READ_WRITE_DELAY_US);
+
+	if (!fu_ccgx_hpi_get_event (device,
+				   hpi_handle,
+				   HPI_REG_SECTION_DEV,
+				   &hpi_event,
+				   HPI_CMD_COMMAND_RESPONSE_TIME_MS,
+				   error)) {
+		g_prefix_error (error, "read flash resp error: ");
+		return FALSE;
+	}
+
+	if (hpi_event != CY_PD_RESP_FLASH_DATA_AVAILABLE) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "read flash resp code error: 0x%x", hpi_event);
+		return FALSE;
+	}
+
+	g_usleep (HPI_CMD_FLASH_READ_WRITE_DELAY_US);
+
+	if (hpi_handle->hpi_addr_byte > 1) {
+		reg_addr = HPI_DEV_REG_FLASH_MEM;
+	} else {
+		reg_addr = CY_PD_REG_BOOTDATA_MEMEORY_ADDR;
+	}
+
+	if (!fu_ccgx_hpi_read_reg(device,hpi_handle, reg_addr, data, size,error)) {
+		g_warning("read data from memory error");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/**
+ * fu_ccgx_hpi_cmd_validate_fw:
+ * @device: #FuDevice
+ * @hpi_handle HPI handle
+ * @fw_index firmeware Index (FW1 or FW2)
+ * @error: a #GError or %NULL
+ *
+ * Validate firmware written in PD I2C Device
+ *
+ * Returns: %TRUE for success
+*/
+gboolean
+fu_ccgx_hpi_cmd_validate_fw (FuDevice *device,
+			     CyHPIHandle *hpi_handle,
+			     guint8 fw_index,
+			     GError **error)
+{
+	CyPDResp hpi_event = 0;
+
+	g_return_val_if_fail (hpi_handle != NULL, FALSE);
+
+	(void)fu_ccgx_hpi_clear_all_event (device,hpi_handle,HPI_CMD_COMMAND_CLEAR_EVENT_TIME_MS,error);
+
+	if (!fu_ccgx_hpi_validate_fw (device, hpi_handle, fw_index,error)) {
+		g_warning("validate fw cmd error");
+		return FALSE;
+	}
+
+	if (!fu_ccgx_hpi_get_event (device, hpi_handle, HPI_REG_SECTION_DEV, &hpi_event, HPI_CMD_COMMAND_RESPONSE_TIME_MS ,error)) {
+		g_warning("validate fw resp error");
+		return FALSE;
+	}
+
+	if (hpi_event != CY_PD_RESP_SUCCESS) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "validate fw resp code error: 0x%x", 
+			     hpi_event);
+	}
+	return TRUE;
+}
+
+/**
+ * fu_ccgx_hpi_cmd_reset_device:
+ * @device: #FuDevice
+ * @hpi_handle HPI handle
+ * @error: a #GError or %NULL
+ *
+ * Reset PD I2C Device
+ *
+ * Returns: %TRUE for success
+*/
+gboolean
+fu_ccgx_hpi_cmd_reset_device (FuDevice *device, CyHPIHandle *hpi_handle, GError **error)
+{
+	g_return_val_if_fail (hpi_handle != NULL, FALSE);
+
+	(void)fu_ccgx_hpi_clear_all_event (device, hpi_handle, HPI_CMD_COMMAND_CLEAR_EVENT_TIME_MS,	error);
+
+	if (!fu_ccgx_hpi_reset_device(device,hpi_handle,error)) {
+		g_warning("reset device cmd error");
```

Comment:
should use g_prefix_error and needs to return false

Reply:
OK. I will use g_prefix_error and add return FALSE.

Reply:
I fixed it.